File: integrations/open_webui/opticparse_tool.py
"""
title: OpticParse & PhishVision Intelligence Tools
author: parastejpal987
author_url: https://github.com/parastejpal987-cmyk/opticparse
git_url: https://github.com/parastejpal987-cmyk/opticparse
description: Autonomous multimodal vision web scraper & real-time cybersecurity phishing detection scanner. Bypasses dynamic JS, bot challenges, and identifies zero-day impersonation threats.
version: 1.0.0
licence: MIT
requirements: httpx, pydantic
"""

import json
from typing import Any, Callable, Dict, Optional
from urllib.parse import urlparse
import httpx
from pydantic import BaseModel, Field


class Tools:
    class Valves(BaseModel):
        OPTICPARSE_API_KEY: str = Field(
            default="",
            description="Optional OpticParse API key for higher rate limits. Free trial quota available by default.",
        )
        PORTAL_URL: str = Field(
            default="https://opticparse-api.onrender.com",
            description="OpticParse API Portal URL.",
        )
        TIMEOUT_SECONDS: int = Field(
            default=45,
            description="Network timeout in seconds for rendering and scanning.",
        )

    def __init__(self):
        self.valves = self.Valves()

    def _normalize_and_validate_url(self, target: str) -> str:
        clean = target.strip()
        if not clean:
            raise ValueError("Target URL or domain cannot be empty.")

        # If a scheme is explicitly provided, verify it is strictly http or https
        parsed_initial = urlparse(clean)
        if parsed_initial.scheme and parsed_initial.scheme not in ("http", "https"):
            raise ValueError(f"Invalid or unsupported URL scheme: {parsed_initial.scheme}")

        if not clean.startswith(("http://", "https://")):
            clean = f"https://{clean}"

        parsed = urlparse(clean)
        if not (parsed.scheme in ("http", "https") and parsed.netloc):
            raise ValueError(f"Invalid target URL or domain: {target}")

        return clean

    def _get_portal_url(self) -> str:
        portal = self.valves.PORTAL_URL.strip().rstrip("/")
        api_key = self.valves.OPTICPARSE_API_KEY.strip()

        # Enforce HTTPS if an API key is configured to protect credentials
        if api_key and portal.startswith("http://"):
            raise ValueError(
                "Insecure HTTP portal URL is not allowed when an API key is configured. Use HTTPS."
            )
        return portal

    def _get_headers(self) -> Dict[str, str]:
        headers = {
            "Content-Type": "application/json",
            "User-Agent": "OpenWebUI-OpticParse-Tool/1.0.0",
        }
        api_key = self.valves.OPTICPARSE_API_KEY.strip()
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"
            headers["X-API-Key"] = api_key
        return headers

    async def opticparse_scrape(
        self,
        url: str,
        query: str = "",
        __event_emitter__: Optional[Callable[[Dict[str, Any]], Any]] = None,
    ) -> str:
        """
        Extract clean, structured web content and markdown from any URL using multimodal computer vision.
        Bypasses Cloudflare Turnstile, heavy JavaScript SPAs, and bot protections.

        :param url: The target website URL to scrape (e.g. https://news.ycombinator.com).
        :param query: Optional specific natural language query or instructions on what to extract.
        :return: Extracted structured markdown or JSON content.
        """
        try:
            target_url = self._normalize_and_validate_url(url)
            portal = self._get_portal_url()
            headers = self._get_headers()

            if __event_emitter__:
                await __event_emitter__(
                    {
                        "type": "status",
                        "data": {
                            "description": f"OpticParse rendering and extracting content from {target_url}...",
                            "done": False,
                        },
                    }
                )

            payload = {
                "url": target_url,
                "query": query or "Extract all primary text, articles, tables, and structured data.",
            }

            async with httpx.AsyncClient(
                timeout=float(self.valves.TIMEOUT_SECONDS),
                follow_redirects=False,
            ) as client:
                endpoint = f"{portal}/api/v1/parse"
                response = await client.post(endpoint, json=payload, headers=headers)

                # Fallback to MCP JSON-RPC call if REST returns 404/405
                if response.status_code in (404, 405):
                    mcp_endpoint = f"{portal}/mcp"
                    rpc_payload = {
                        "jsonrpc": "2.0",
                        "id": 1,
                        "method": "tools/call",
                        "params": {
                            "name": "opticparse_scrape",
                            "arguments": {
                                "target_url": target_url,
                                "extraction_query": query,
                            },
                        },
                    }
                    response = await client.post(mcp_endpoint, json=rpc_payload, headers=headers)

                if response.status_code in (301, 302, 307, 308):
                    return "Error: Gateway returned an unexpected redirect. Request aborted to prevent credential leakage."

                response.raise_for_status()
                data = response.json()

                if __event_emitter__:
                    await __event_emitter__(
                        {
                            "type": "status",
                            "data": {
                                "description": "OpticParse extraction complete.",
                                "done": True,
                            },
                        }
                    )

                if isinstance(data, dict):
                    if "content" in data:
                        return str(data["content"])
                    if "result" in data:
                        return json.dumps(data["result"], indent=2)
                    return json.dumps(data, indent=2)
                return str(data)

        except Exception as e:
            if __event_emitter__:
                await __event_emitter__(
                    {
                        "type": "status",
                        "data": {
                            "description": f"OpticParse error: {str(e)}",
                            "done": True,
                        },
                    }
                )
            return f"Error executing OpticParse: {str(e)}"

    async def phishvision_scan(
        self,
        target: str,
        __event_emitter__: Optional[Callable[[Dict[str, Any]], Any]] = None,
    ) -> str:
        """
        Inspect a URL or domain for zero-day phishing, credential harvesting, brand impersonation, and wallet drainers.
        Scans visual layout, DOM anomalies, SSL/DNS records, and threat intelligence in real time.

        :param target: The website URL or bare domain to inspect (e.g. example.com or https://suspicious-login.xyz).
        :return: Comprehensive security audit report including risk score, verdict, and detected IOCs.
        """
        try:
            target_url = self._normalize_and_validate_url(target)
            portal = self._get_portal_url()
            headers = self._get_headers()

            if __event_emitter__:
                await __event_emitter__(
                    {
                        "type": "status",
                        "data": {
                            "description": f"PhishVision analyzing threat profile for {target_url}...",
                            "done": False,
                        },
                    }
                )

            payload = {"url": target_url}

            async with httpx.AsyncClient(
                timeout=float(self.valves.TIMEOUT_SECONDS),
                follow_redirects=False,
            ) as client:
                endpoint = f"{portal}/phishvision/scan"
                response = await client.post(endpoint, json=payload, headers=headers)

                # Fallback to MCP JSON-RPC call if REST returns 404/405
                if response.status_code in (404, 405):
                    mcp_endpoint = f"{portal}/mcp"
                    rpc_payload = {
                        "jsonrpc": "2.0",
                        "id": 1,
                        "method": "tools/call",
                        "params": {
                            "name": "phishvision_detect",
                            "arguments": {"target_url": target_url},
                        },
                    }
                    response = await client.post(mcp_endpoint, json=rpc_payload, headers=headers)

                if response.status_code in (301, 302, 307, 308):
                    return "Error: Gateway returned an unexpected redirect. Scan aborted to protect integrity."

                response.raise_for_status()
                data = response.json()

                if __event_emitter__:
                    await __event_emitter__(
                        {
                            "type": "status",
                            "data": {
                                "description": "PhishVision threat scan complete.",
                                "done": True,
                            },
                        }
                    )

                if isinstance(data, dict):
                    return json.dumps(data, indent=2)
                return str(data)

        except Exception as e:
            if __event_emitter__:
                await __event_emitter__(
                    {
                        "type": "status",
                        "data": {
                            "description": f"PhishVision error: {str(e)}",
                            "done": True,
                        },
                    }
                )
            return f"Error executing PhishVision scan: {str(e)}"

File: integrations/flowise/OpticParse_Tool.ts
/**
 * OpticParse Flowise Custom Tool Node
 * Autonomous Multimodal Vision Web Scraper for Flowise Chatflows & Agents
 */

export interface INodeParams {
    label: string;
    name: string;
    type: string;
    default?: any;
    optional?: boolean;
    description?: string;
    placeholder?: string;
}

export interface INode {
    label: string;
    name: string;
    version: number;
    description: string;
    type: string;
    icon: string;
    category: string;
    baseClasses: string[];
    inputs: INodeParams[];
    run?(nodeData: any, input: string, options?: any): Promise<string>;
}

export function normalizeTargetUrl(target: string): string {
    const clean = (target || '').trim();
    if (!clean) {
        throw new Error('Target URL or domain cannot be empty.');
    }

    // Check scheme if provided
    const schemeMatch = clean.match(/^([a-zA-Z][a-zA-Z0-9+.-]*):/);
    if (schemeMatch) {
        const scheme = schemeMatch[1].toLowerCase();
        if (scheme !== 'http' && scheme !== 'https') {
            throw new Error(`Invalid or unsupported URL scheme: ${scheme}`);
        }
    }

    const normalized = clean.startsWith('http://') || clean.startsWith('https://')
        ? clean
        : `https://${clean}`;

    try {
        const parsed = new URL(normalized);
        if (!parsed.hostname) {
            throw new Error(`Invalid target URL or domain: ${target}`);
        }
    } catch {
        throw new Error(`Invalid target URL or domain: ${target}`);
    }

    return normalized;
}

export function resolvePortalUrl(portalUrl: string, apiKey: string): string {
    const portal = (portalUrl || 'https://opticparse-api.onrender.com').trim().replace(/\/+$/, '');
    if (apiKey && portal.startsWith('http://')) {
        throw new Error('Insecure HTTP portal URL is not allowed when an API key is configured. Use HTTPS.');
    }
    return portal;
}

export class OpticParse_Tool implements INode {
    label = 'OpticParse Web Scraper';
    name = 'opticParseTool';
    version = 1.0;
    type = 'OpticParseTool';
    icon = 'opticparse.svg';
    category = 'Tools';
    description = 'Autonomous vision scraper that renders modern Single Page Apps, solves dynamic JavaScript, and bypasses Cloudflare bot protection.';
    baseClasses = ['Tool', 'StructuredTool'];

    inputs: INodeParams[] = [
        {
            label: 'Target URL',
            name: 'url',
            type: 'string',
            placeholder: 'https://news.ycombinator.com',
            description: 'The website URL to scrape.'
        },
        {
            label: 'Extraction Query',
            name: 'query',
            type: 'string',
            optional: true,
            placeholder: 'Extract article headlines and links in markdown format',
            description: 'Natural language instruction defining what elements to extract.'
        },
        {
            label: 'OpticParse API Key',
            name: 'apiKey',
            type: 'password',
            optional: true,
            description: 'Optional API key for higher rate limits. Free trial quota available by default.'
        },
        {
            label: 'Portal URL',
            name: 'portalUrl',
            type: 'string',
            optional: true,
            default: 'https://opticparse-api.onrender.com',
            description: 'OpticParse gateway portal URL.'
        }
    ];

    async run(nodeData: any, input: string): Promise<string> {
        const rawUrl = nodeData?.inputs?.url || input;
        const query = nodeData?.inputs?.query || 'Extract all primary text, articles, tables, and structured data.';
        const apiKey = (nodeData?.inputs?.apiKey || '').trim();
        const portalUrl = resolvePortalUrl(nodeData?.inputs?.portalUrl, apiKey);
        const targetUrl = normalizeTargetUrl(rawUrl);

        const headers: Record<string, string> = {
            'Content-Type': 'application/json',
            'User-Agent': 'Flowise-OpticParse-Tool/1.0.0'
        };
        if (apiKey) {
            headers['Authorization'] = `Bearer ${apiKey}`;
            headers['X-API-Key'] = apiKey;
        }

        const endpoint = `${portalUrl}/api/v1/parse`;
        const payload = { url: targetUrl, query };

        const response = await fetch(endpoint, {
            method: 'POST',
            headers,
            body: JSON.stringify(payload),
            redirect: 'manual'
        });

        if (response.status >= 300 && response.status < 400) {
            return 'Error: Gateway returned an unexpected redirect. Request aborted to prevent credential leakage.';
        }

        if (response.status === 404 || response.status === 405) {
            // Fallback to JSON-RPC MCP endpoint
            const mcpEndpoint = `${portalUrl}/mcp`;
            const rpcPayload = {
                jsonrpc: '2.0',
                id: 1,
                method: 'tools/call',
                params: {
                    name: 'opticparse_scrape',
                    arguments: { target_url: targetUrl, extraction_query: query }
                }
            };
            const rpcResponse = await fetch(mcpEndpoint, {
                method: 'POST',
                headers,
                body: JSON.stringify(rpcPayload),
                redirect: 'manual'
            });
            if (rpcResponse.status >= 300 && rpcResponse.status < 400) {
                return 'Error: MCP Gateway returned an unexpected redirect.';
            }
            if (!rpcResponse.ok) {
                return `Error from OpticParse MCP: HTTP ${rpcResponse.status}`;
            }
            const rpcData = await rpcResponse.json();
            return typeof rpcData?.result === 'object' ? JSON.stringify(rpcData.result, null, 2) : String(rpcData?.result || '');
        }

        if (!response.ok) {
            return `Error from OpticParse: HTTP ${response.status} - ${await response.text()}`;
        }

        const data = await response.json();
        if (data && typeof data === 'object') {
            if (data.content) return String(data.content);
            if (data.result) return JSON.stringify(data.result, null, 2);
            return JSON.stringify(data, null, 2);
        }
        return String(data);
    }
}

export default new OpticParse_Tool();
